snowpark_style_guide.md

```diff
@@ -0,0 +1,325 @@
+```python
+# select by attribute
+df = df.select(df.col_a, df.col_b)
+
+# select by string
+df = df.select("col_a", "col_b")
+
+# select by using snowflake.snowpark.functions.col function
+df = df.select(col("col_a"), col("col_b"))
+
+# select by Dataframe indexing
+df = df.select(df["col_a"], df["col_b"])
+
+# select by using DataFrame.col function
+df = df.select(df.col("col_a"), df.col("col_b"))
+```
+
+- The attribute approach is shorter and simpler, however, it cannot tackle column names containing
+special characters.
+- The string approach has the same benefits as the first one, but also supporting names with special characters.
+- The `col` function approach will wrap the col as a `Column` object. It's the best of usage when operations on `Column` like aliasing the column name are needed.
+- The indexing and `DataFrame.col` approaches offer a way to disambiguate among multiple `DataFrame`.
+Suppose there are two `DataFrame` with same column name `col_a` and a `join` operation is wanted, it could be accomplished by:
+    ```python
+    df = df1.join(df2, df2["col_a"] == df1["col_a"])
+    ```
+
+## Aliasing Column Name
+
+Aliasing is typically useful for providing meaningful names for columns.
+Sometimes the column names in the database is either composed of long words for the purpose of
+being descriptive or too short lacking enough information. Keeping those long column names while manipulating `DataFrame`
+could somehow reduce maintainability and readability such that aliasing is wanted.
+
+There are two ways to do aliasing in Snowpark : either `DataFrame.with_column_renamed` or `Column.alias`
+suffices the purpose. `with_column_renamed` separates column selection and renaming.
+In contrast, `Column.alias` brings the benefit
+of locality as well as makes code more compact which further improves readability and conciseness.
+
+```python
+# with_column_renamed
+order = order.select(
+    'restaurant_id',
+    'order_id',
+    'order_type',
+    'order_customer_id',
+    'price',
+    'time'
+)
+order = order.with_column_renamed('order_customer_id', 'customer_id')
+order = order.with_column_renamed('price', 'order_price')
+order = order.with_column_renamed('time', 'order_time')
+
+# alias
+order = order.select(
+    'restaurant_id',
+    'order_id',
+    'order_type',
+    col('order_customer_id').alias('customer_id'),
+    col('cost').alias('order_cost'),
+    col('time').alias('order_time')
+)
+```
+
+## Performant SQL Generation
+
+Operations upon `DataFrame` will ultimately get turned into SQL statements and send to Snowflake for execution.
+Non-optimized SQL statements would take longer time to process and execute and sometimes the service even times out.
+Snowpark has implemented a SQL simplifier which is turned on by default, and it is build upon certain rules.
+To maximize the effectiveness of the simplifier, we propose the following guidelines based on our practices:
+
+<TO BE FILLED>
```

Comment:
@sfc-gh-yixie I might not have the same deep understand as yours on this part, is there something you'd like to call out?